### DISCUSSION TITLE
How do I Integrate with Visual Studio Code (linux)

### DISCUSSION THREAD

**Discussion by mohamedsayed18** [microsoft/vcpkg#17814] [Ideas]
I following the readme to install vcpkg
when I run `./vcpkg integrate install`
output
```
Applied user-wide integration for this vcpkg root.
CMake projects should use: "-DCMAKE_TOOLCHAIN_FILE=/home/mohamed/vcpkg/scripts/buildsystems/vcpkg.cmake"
```
I tried these commands
```
mkdir build
cd build
cmake /home/mohamed/vcpkg/toolsrc/ -DCMAKE_TOOLCHAIN_FILE=/home/mohamed/vcpkg/scripts/buildsystems/vcpkg.cmake
```
the last line of output `Build files have been written to: /home/mohamed/vcpkg/build`
but when I run the `./vcpkg integrate install` again it gives the same error .
I'm totally new to cmake. am I using the right cmakelists.txt which is inside the toolsrc
Environment:
* ubuntu 18.04

**Comment by syys96** [comment]
Can you compile your project with make after cmake command? the message "Build files have been written to: /home/mohamed/vcpkg/build" means  that you have successfully configured it

**Comment by mohamedsayed18** [comment]
I cd into the build directory and try `make`
output:
```
Scanning dependencies of target vcpkglib
[  1%] Building CXX object CMakeFiles/vcpkglib.dir/src/vcpkg/archives.cpp.o
[  2%] Building CXX object CMakeFiles/vcpkglib.dir/src/vcpkg/base/checks.cpp.o
[  3%] Building CXX object CMakeFiles/vcpkglib.dir/src/vcpkg/base/chrono.cpp.o
[  4%] Building CXX object CMakeFiles/vcpkglib.dir/src/vcpkg/base/cofffilereader.cpp.o
[  5%] Building CXX object CMakeFiles/vcpkglib.dir/src/vcpkg/base/downloads.cpp.o
[  6%] Building CXX object CMakeFiles/vcpkglib.dir/src/vcpkg/base/enums.cpp.o
[  7%] Building CXX object CMakeFiles/vcpkglib.dir/src/vcpkg/base/files.cpp.o
[  8%] Building CXX object CMakeFiles/vcpkglib.dir/src/vcpkg/base/hash.cpp.o
[  9%] Building CXX object CMakeFiles/vcpkglib.dir/src/vcpkg/base/json.cpp.o
[ 10%] Building CXX object CMakeFiles/vcpkglib.dir/src/vcpkg/base/machinetype.cpp.o
[ 11%] Building CXX object CMakeFiles/vcpkglib.dir/src/vcpkg/base/parse.cpp.o
[ 12%] Building CXX object CMakeFiles/vcpkglib.dir/src/vcpkg/base/strings.cpp.o
[ 13%] Building CXX object CMakeFiles/vcpkglib.dir/src/vcpkg/base/stringview.cpp.o
[ 14%] Building CXX object CMakeFiles/vcpkglib.dir/src/vcpkg/base/system.cpp.o
[ 15%] Building CXX object CMakeFiles/vcpkglib.dir/src/vcpkg/base/system.print.cpp.o
[ 16%] Building CXX object CMakeFiles/vcpkglib.dir/src/vcpkg/base/system.process.cpp.o
[ 17%] Building CXX object CMakeFiles/vcpkglib.dir/src/vcpkg/base/uint128.cpp.o
[ 18%] Building CXX object CMakeFiles/vcpkglib.dir/src/vcpkg/base/unicode.cpp.o
[ 19%] Building CXX object CMakeFiles/vcpkglib.dir/src/vcpkg/binarycaching.cpp.o
[ 20%] Building CXX object CMakeFiles/vcpkglib.dir/src/vcpkg/binaryparagraph.cpp.o
In file included from /home/mohamed/vcpkg/toolsrc/src/vcpkg/binaryparagraph.cpp:7:0:
/home/mohamed/vcpkg/toolsrc/include/vcpkg/binaryparagraph.h: In constructor ‘vcpkg::BinaryParagraph::BinaryParagraph(const vcpkg::SourceParagraph&, vcpkg::Triplet, const string&, const std::vector<vcpkg::FeatureSpec>&)’:
/home/mohamed/vcpkg/toolsrc/include/vcpkg/binaryparagraph.h:41:14: error: ‘vcpkg::BinaryParagraph::type’ will be initialized after [-Werror=reorder]
         Type type;
              ^~~~
/home/mohamed/vcpkg/toolsrc/include/vcpkg/binaryparagraph.h:38:34: error:   ‘std::vector<std::__cxx11::basic_string<char> > vcpkg::BinaryParagraph::default_features’ [-Werror=reorder]
         std::vector<std::string> default_features;
                                  ^~~~~~~~~~~~~~~~
/home/mohamed/vcpkg/toolsrc/src/vcpkg/binaryparagraph.cpp:84:5: error:   when initialized here [-Werror=reorder]
     BinaryParagraph::BinaryParagraph(const SourceParagraph& spgh,
     ^~~~~~~~~~~~~~~
In file included from /home/mohamed/vcpkg/toolsrc/src/vcpkg/binaryparagraph.cpp:7:0:
/home/mohamed/vcpkg/toolsrc/include/vcpkg/binaryparagraph.h: In constructor ‘vcpkg::BinaryParagraph::BinaryParagraph(const vcpkg::SourceParagraph&, const vcpkg::FeatureParagraph&, vcpkg::Triplet, const std::vector<vcpkg::FeatureSpec>&)’:
/home/mohamed/vcpkg/toolsrc/include/vcpkg/binaryparagraph.h:41:14: error: ‘vcpkg::BinaryParagraph::type’ will be initialized after [-Werror=reorder]
         Type type;
              ^~~~
/home/mohamed/vcpkg/toolsrc/include/vcpkg/binaryparagraph.h:38:34: error:   ‘std::vector<std::__cxx11::basic_string<char> > vcpkg::BinaryParagraph::default_features’ [-Werror=reorder]
         std::vector<std::string> default_features;
                                  ^~~~~~~~~~~~~~~~
/home/mohamed/vcpkg/toolsrc/src/vcpkg/binaryparagraph.cpp:100:5: error:   when initialized here [-Werror=reorder]
     BinaryParagraph::BinaryParagraph(const SourceParagraph& spgh,
     ^~~~~~~~~~~~~~~
At global scope:
cc1plus: error: unrecognized command line option ‘-Wno-redundant-move’ [-Werror]
cc1plus: all warnings being treated as errors
CMakeFiles/vcpkglib.dir/build.make:329: recipe for target 'CMakeFiles/vcpkglib.dir/src/vcpkg/binaryparagraph.cpp.o' failed
make[2]: *** [CMakeFiles/vcpkglib.dir/src/vcpkg/binaryparagraph.cpp.o] Error 1
CMakeFiles/Makefile2:126: recipe for target 'CMakeFiles/vcpkglib.dir/all' failed
make[1]: *** [CMakeFiles/vcpkglib.dir/all] Error 2
Makefile:114: recipe for target 'all' failed
make: *** [all] Error 2```

**Comment by JackBoosY** [comment]
#10297.

@strega-nil Could you please take a look?

Thanks.

**Comment by syys96** [comment]
I got it! The message
    ""Applied user-wide integration for this vcpkg root. CMake projects should use: 
    "DCMAKE_TOOLCHAIN_FILE=/home/mohamed/vcpkg/scripts/buildsystems/vcpkg.cmake"
indicates you have completed the install of vcpkg. It's just a useful information for cmake user. So you don't need to build the vcpkg again, the "./bootstrap-vcpkg.sh" has done it.
Now you can  try to install any packages with " vcpkg install foo"

**Comment by syys96** [comment]
Do't forget to add vcpkg directory to the PATH variable

**Comment by mohamedsayed18** [comment]
I realized that the "CMake projects should use: "-DCMAKE_TOOLCHAIN_FILE" message is not an error. when I try to `#include <box2d/box2d.h>` in vscode. I got this error:
```
fatal error: box2d/box2d.h: No such file or directory
#include <box2d/box2d.h>
```
even the vscode autocomplete the library when I include it

**Comment by mohamedsayed18** [comment]
> Do't forget to add vcpkg directory to the PATH variable

can you tell me where exactly the PATH variable is it in the settings.json?
sorry, I'm totally new to make and package manager

**Comment by syys96** [comment]
It's a system variable of linux configured in the file ~/.bashrc, you can open it with vim or other text editor and add the line:
     export PATH=/home/mohamed/vcpkg:$PATH
then save it and run source ~/.bashrc to activate it.

**Comment by syys96** [comment]
**test_cpp.cpp**
#include <box2d/box2d.h>
#include < iostream>
using namespace std;
int main()
{
    cout << "hello, world!" << endl;
}
**CMakeLists.txt**
cmake_minimum_required(VERSION 3.12)
project(test_cpp)
find_package(unofficial-box2d CONFIG REQUIRED)
add_executable(test_cpp test_cpp.cpp)
target_link_libraries(test_cpp PRIVATE unofficial::box2d::box2d)
**Output**
$ cmake .. -DCMAKE_TOOLCHAIN_FILE=/home/wumian/Documents/soft_ware/vcpkg/scripts/
buildsystems/vcpkg.cmake 
-- Configuring done
-- Generating done
-- Build files have been written to: /home/wumian/Documents/vs_pro/mybuild
$ make
[100%] Built target test_cpp
$ ./test_cpp 
hello, world!

**Comment by syys96** [comment]
I edit the source file in vscode and build in it's inner terminal successfully. But I can't use the **cmake configure** and **cmake build** operation in vscode to build the example. It's very strange

**Comment by strega-nil** [comment]
@syys96 you need to add to your `settings.json` your `CMAKE_TOOLCHAIN_FILE`. You can model it on [strega-nil/vcpkg-example](https://github.com/strega-nil/vcpkg-example) (ignore the `vcpkg.json` stuff)

**Comment by zipzit** [comment]
I'm having essentially the same issue.  Do I add my stuff here, or pull a new issue?  
```
$ hostnamectl 
Operating System: Ubuntu 20.04 LTS
Kernel: Linux 5.4.0-31-generic
Architecture: x86-64
```
I carefully went thru the installation of vcpkg.  

```
$  ./vcpkg integrate install 
Applied user-wide integration for this vcpkg root.
CMake projects should use: "-DCMAKE_TOOLCHAIN_FILE=/home/lbc/vcpkg/scripts/buildsystems/vcpkg.cmake"

$ ./vcpkg list
fmt:x64-linux             6.2.0            Formatting library for C++. It can be used as a ...

$ echo $PATH
/home/lbc/vcpkg:/home/lbc/.nvm/versions/node/v12.16.1/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
```

For my testing,  I'm using [FMT](https://fmt.dev/latest/index.html)
Test code:

```
#include <fmt/core.h>      <--- total fail, Cannot open source file, C/C++(1696)
int main() {
  fmt::print("The answer is {}.", 42);
}
```
I'm using VS Code, version 1.45.1   with C/C++ for Visual Studio Code Extension (0.28.1) installed.  As I understand it, from this [link](https://devblogs.microsoft.com/cppblog/visual-studio-code-cc-extension-may-2018-update-intellisense-configuration-just-got-so-much-easier/) that Visual Studio Code C/C++ extension May 2018 Update includes Auto-detection of libraries installed by Vcpkg for IntelliSense.  

I've tried both sudo and without for `./vcpkg integrate install`   All the responses indicate everything is whole, yet, its a total fail finding the import in VS code.  Any ideas (or tests to run?)  

I'm wondering if my use case is right:  

- Full functionality of C++
- with modern libraries
- while using VS Code
- in Linux.

Many thanks for the support here...

**Comment by syys96** [comment]
Can you build the test in terminal instead of vscode? Or do you add the CMAKE_TOOLCHAIN_FILE to setting.json under your workplace?

**Comment by syys96** [comment]
**CMakeLists.txt**
cmake_minimum_required(VERSION 3.12)
set(CMAKE_TOOLCHAIN_FILE /home/lbc/vcpkg/scripts/buildsystems/vcpkg.cmake)
project(test_cpp)
find_package(fmt CONFIG REQUIRED)
add_executable(test_cpp main.cpp)
target_link_libraries(test_cpp PRIVATE fmt::fmt fmt::fmt-header-only)

Then you can use cmake and make to build the test, the progress has successfully passed on my machine

**Comment by zipzit** [comment]
@syys96 You are my hero!  That totally nailed it.  I had some sloppiness within my CMakeLists.txt file, at the `find_package()` command and `target_link_libraries()` command.  

I will say, I don't understand the linker line of code `target_link_libraries(CmakeVcpkg PRIVATE fmt::fmt fmt::fmt-header-only)`.  What's up with `fmt::fmt fmt::fmt-header-only` ?   Where did that format come from?  Is that a standard format, or just some quirky stuff from the **fmt** library?

Note, I've been trying to really understand how to correctly use C++ in Linux with VS Code, with a real deugger.  In fact i created a decent repo on some expected use cases... In fact, would this be a potential start towards creating a decent beginners tutorial on using VCPKG with VS Code?    

https://github.com/zipzit/vscode_cpp_projects    

I will say, the **fmt** library is sort of obtuse. That's not a library I would every be likely to use.  I'm thinking of adding a simple graphics or charting library via VCPKG would be a nice touch.  

any feedback?  And syys96, thank you for your assistance here!

**Comment by strega-nil** [comment]
@zipzit yeah, that's some oddity with the fmt library. Basically, you can use it either ["header-only"], or compile it separately. Since you're using vcpkg already to take care of that, there's no reason not to just use `fmt::fmt`.

["header-only"]: https://en.wikipedia.org/wiki/Header-only

**Comment by zipzit** [comment]
I'm looking at the topic question here: "How do I Integrate with Visual Studio Code (linux)"
And for pure virgins its a multipart answer., [reference](https://github.com/zipzit/vscode_cpp_projects,)
1) Figure out how to use VScode for C++ (see [1_cpp_helloworld.](https://github.com/zipzit/vscode_cpp_projects/tree/master/1_cpp_helloworld) )
2) Figure out how to use CMake with VScode for C++ (see [2_cmake_helloworld](https://github.com/zipzit/vscode_cpp_projects/tree/master/2_cmake_helloworld))
3) Figure out how to use vcpkg with VScode with C++ (see [5_cmake_vcpkg](https://github.com/zipzit/vscode_cpp_projects/tree/master/5_cmake_vcpkg) )

I'm wondering if it would be appropriate to do a pull request to add this stuff as a VScode how-to for this repo?
